### PR TITLE
Install spec implementation

### PR DESCRIPTION
**Pull request by m-blaha**
New high level API for installing any artifact using the same simple call goal.add_install(spec).

The spec here can be
- a repository package specification matched against NEVRA, provides, and file provides, e.g. "wget", "vim-*"
- a local rpm file (/the/location/package.rpm)
- a URL of remote rpm package
- by using '@' prefix also a group can be specified - @rpm-development-tools
- TODO: once installation of environmental groups and modules is  implemented, also these artifact will be able to be installed

### CODE REVIEW

File: libdnf/repo/repo_sack.cpp [hunk 1]
```diff
@@ -84,6 +88,93 @@ RepoWeakPtr RepoSack::get_cmdline_repo() {
...
+
+    auto cmdline_repo = get_cmdline_repo();
+
+    // Ensure that download location for command line repository exists
+    std::filesystem::path cmd_repo_pkgs_dir{cmdline_repo->get_cachedir()};
+    cmd_repo_pkgs_dir /= CACHE_PACKAGES_DIR;
+    std::filesystem::create_directories(cmd_repo_pkgs_dir);
+
+    // map remote URLs to local destination files
+    std::map<std::string, std::filesystem::path> url_to_path;
```
Comment (jrohel) [inline libdnf/repo/repo_sack.cpp:117]: Minor: What about use `std::string_view` to save some memory?
`std::map<std::string_view, std::filesystem::path> url_to_path;`
and use `std::string{url}` on line 153.

File: libdnf/repo/repo_sack.cpp [hunk 2]
```diff
@@ -84,6 +88,93 @@ RepoWeakPtr RepoSack::get_cmdline_repo() {
...
+    for (const auto & url : rpm_urls) {
+        if (url_to_path.find(url) != url_to_path.end()) {
```
Comment (jrohel) [inline libdnf/repo/repo_sack.cpp:119]: Minor: How about using `if (url_to_path.contains(url)) {`?

File: libdnf/base/goal.cpp [hunk 3]
```diff
@@ -1438,6 +1495,32 @@ GoalProblem Goal::Impl::add_reason_change_to_goal(
...
+    // fill the command line repo with paths to rpm files
+    std::vector<std::string> paths;
+    for (const auto & [action, path, settings] : rpm_filepaths) {
+        paths.emplace_back(path);
+    }
+    auto cmdline_packages = base->get_repo_sack()->add_cmdline_packages(paths);
+
+    // add newly created packages to the goal
+    for (const auto & [action, path, settings] : rpm_filepaths) {
+        printf("XXX path: %s\n", path.c_str());
```
Comment (jrohel) [inline libdnf/base/goal.cpp:1512]: Forgotten debug print?

File: include/libdnf/base/goal.hpp [hunk 4]
```diff
@@ -40,6 +40,39 @@ class Goal {
...
+
+    /// High level API for an artifact downgrade. See `add_install()` for details.
+    /// @param spec      A string with upgrade spec
+    /// @param settings  A structure to override default goal settings.
+    void add_downgrade(const std::string & spec, const libdnf::GoalJobSettings & settings = libdnf::GoalJobSettings());
+
+    /// High level API for an artifact reinstall. See `add_install()` for details.
+    /// @param spec      A string with reinstall spec
+    /// @param settings  A structure to override default goal settings.
+    void add_reinstall(const std::string & spec, const libdnf::GoalJobSettings & settings = libdnf::GoalJobSettings());
```
[the remark this answers is not in the document]
Comment (m-blaha) [reply include/libdnf/base/goal.hpp:69]: There was an assert preventing libdnf to perform some actions on groups. Actually currently only INSTALL and REMOVE actions are supported.
I've added commits to change this assert and a resolve log is written instead.

File: libdnf/base/goal.cpp (hunk 3 above)
Comment (m-blaha) [reply libdnf/base/goal.cpp]: Yeah. Fixed.

File: libdnf/repo/repo_sack.cpp (hunk 1 above)
Comment (m-blaha) [reply libdnf/repo/repo_sack.cpp]: Done.

File: libdnf/repo/repo_sack.cpp (hunk 2 above)
Comment (m-blaha) [reply libdnf/repo/repo_sack.cpp]: Done.

File: libdnf/repo/repo_sack.cpp [hunk 5]
```diff
@@ -84,6 +88,93 @@ RepoWeakPtr RepoSack::get_cmdline_repo() {
...
+
+        // fill the command line repo with downloaded URLs
+        for (const auto & [url, path] : url_to_path) {
+            path_to_package.emplace(url, cmdline_repo->add_rpm_package(path.string(), true));
+        }
+    }
+
+    // fill the command line repo with local files
+    for (const auto & path : rpm_filepaths) {
+        if (path_to_package.find(path) == path_to_package.end()) {
```
[the remark this answers is not in the document]
Comment (jrohel) [reply libdnf/repo/repo_sack.cpp]: > But the `cmdline_repo->add_rpm_package(path, true)` expression will still be called, won't it?

Yes. You're right. So you can use `if (!package_path.contains(path)) {`.

**Comment by m-blaha** [comment]
Hm, some CI tests started failing. I need to investigate it.

**Comment by jrohel** [comment]
Thank you for the PR and cooperation.

**Pull request merged**